File: Java/LeetCode/Reverse Vowels of a String.java
// https://leetcode.com/problems/reverse-vowels-of-a-string

class Solution {

    final Set<Character> VOWELS = Set.of('a','e','i','o','u','A','E','I','O','U');

    public String reverseVowels(String s) {
        Stack<Character> vowels = new Stack<>();

        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (VOWELS.contains(c)) {
                vowels.push(c);
            }
        }

        StringBuilder output = new StringBuilder();

        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (VOWELS.contains(c)) {
                output.append(vowels.pop());
            } else {
                output.append(c);
            }
        }

        return output.toString();
    }

}

File: JavaScript/LeetCode/Reverse Vowels of a String.ts
// https://leetcode.com/problems/reverse-vowels-of-a-string

function reverseVowels(s: string): string {
    let vowels = [];

    for (const c of s) {
        if (VOWELS.has(c)) {
            vowels.push(c);
        }
    }

    let output = "";

    for (const c of s) {
        if (VOWELS.has(c)) {
            output += vowels.pop();
        } else {
            output += c;
        }
    }

    return output;
};

const VOWELS = new Set([
    'a','e','i','o','u','A','E','I','O','U'
]);
